<div class="inner_cell"><div class="ctb_hideshow"><div class="celltoolbar"></div></div><div class="input_area" aria-label="Edit Markup Text here"><div class="CodeMirror cm-s-default CodeMirror-wrap"><div style="overflow: hidden; position: relative; width: 3px; height: 0px; top: 5.60004px; left: 38px;"><textarea autocorrect="off" autocapitalize="off" spellcheck="false" tabindex="0" style="position: absolute; bottom: -1em; padding: 0px; width: 1000px; height: 1em; outline: none;"></textarea></div><div class="CodeMirror-vscrollbar" tabindex="-1" cm-not-content="true" style="bottom: 0px;"><div style="min-width: 1px; height: 0px;"></div></div><div class="CodeMirror-hscrollbar" tabindex="-1" cm-not-content="true"><div style="height: 100%; min-height: 1px; width: 0px;"></div></div><div class="CodeMirror-scrollbar-filler" cm-not-content="true"></div><div class="CodeMirror-gutter-filler" cm-not-content="true"></div><div class="CodeMirror-scroll" tabindex="-1"><div class="CodeMirror-sizer" style="margin-left: 34px; padding-right: 0px; padding-bottom: 0px; margin-bottom: -13px; border-right-width: 17px; min-height: 300px;"><div style="position: relative; top: 0px;"><div class="CodeMirror-lines" role="presentation"><div role="presentation" style="position: relative; outline: none;"><div class="CodeMirror-measure"><div class="CodeMirror-linenumber CodeMirror-gutter-elt"><div>5</div></div></div><div class="CodeMirror-measure"></div><div style="position: relative; z-index: 1;"></div><div class="CodeMirror-cursors"><div class="CodeMirror-cursor" style="left: 4px; top: 0px; height: 20px;">&nbsp;</div></div><div class="CodeMirror-code" role="presentation" style=""><div style="position: relative;"><div class="CodeMirror-gutter-wrapper" style="left: -34px;"><div class="CodeMirror-linenumber CodeMirror-gutter-elt" style="left: 0px; width: 21px;">1</div></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span class="cm-header cm-header-2">## Getting Started</span></span></pre></div><div style="position: relative;"><div class="CodeMirror-gutter-wrapper" style="left: -34px;"><div class="CodeMirror-linenumber CodeMirror-gutter-elt" style="left: 0px; width: 21px;">2</div></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="">​</span></span></pre></div><div style="position: relative;"><div class="CodeMirror-gutter-wrapper" style="left: -34px;"><div class="CodeMirror-linenumber CodeMirror-gutter-elt" style="left: 0px; width: 21px;">3</div></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">In this project, you will employ several supervised algorithms of your choice to accurately model individuals' income using data collected from the 1994 U.S. Census. You will then choose the best candidate algorithm from preliminary results and further optimize this algorithm to best model the data. Your goal with this implementation is to construct a model that accurately predicts whether an individual makes more than <span class="cm-delimit cm-delimit-open">$</span><span class="cm-atom">50</span>,000. This sort of task can arise in a non-profit setting, where organizations survive on donations.  Understanding an individual's income can help a non-profit better understand how large of a donation to request, or whether or not they should reach out to begin with.  While it can be difficult to determine an individual's general income bracket directly from public sources, we can (as we will see) infer this value from other publically available features. </span></pre></div><div style="position: relative;"><div class="CodeMirror-gutter-wrapper" style="left: -34px;"><div class="CodeMirror-linenumber CodeMirror-gutter-elt" style="left: 0px; width: 21px;">4</div></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;"><span cm-text="">​</span></span></pre></div><div style="position: relative;"><div class="CodeMirror-gutter-wrapper" style="left: -34px;"><div class="CodeMirror-linenumber CodeMirror-gutter-elt" style="left: 0px; width: 21px;">5</div></div><pre class=" CodeMirror-line " role="presentation"><span role="presentation" style="padding-right: 0.1px;">The dataset for this project originates from the <span class="cm-bracket">[</span>UCI Machine Learning Repository<span class="cm-bracket">]</span>(https://archive.ics.uci.edu/ml/datasets/Census+Income). The datset was donated by Ron Kohavi and Barry Becker, after being published in the article _"Scaling Up the Accuracy of Naive-Bayes Classifiers: A Decision-Tree Hybrid"_. You can find the article by Ron Kohavi <span class="cm-bracket">[</span>online<span class="cm-bracket">]</span>(https://www.aaai.org/Papers/KDD/1996/KDD96-033.pdf). The data we investigate here consists of small changes to the original dataset, such as removing the `'fnlwgt'` feature and records with missing or ill-formatted entries.</span></pre></div></div></div></div></div></div><div style="position: absolute; height: 17px; width: 1px; border-bottom: 0px solid transparent; top: 300px;"></div><div class="CodeMirror-gutters" style="left: 0px; height: 317px;"><div class="CodeMirror-gutter CodeMirror-linenumbers" style="width: 33px;"></div></div></div></div></div><div class="text_cell_render rendered_html" tabindex="-1"><h2 id="Getting-Started">Getting Started<a class="anchor-link" href="#Getting-Started">¶</a></h2>
<p>In this project, you will employ several supervised algorithms of your choice to accurately model individuals' income using data collected from the 1994 U.S. Census. You will then choose the best candidate algorithm from preliminary results and further optimize this algorithm to best model the data. Your goal with this implementation is to construct a model that accurately predicts whether an individual makes more than $50,000. This sort of task can arise in a non-profit setting, where organizations survive on donations.  Understanding an individual's income can help a non-profit better understand how large of a donation to request, or whether or not they should reach out to begin with.  While it can be difficult to determine an individual's general income bracket directly from public sources, we can (as we will see) infer this value from other publically available features. </p>
<p>The dataset for this project originates from the <a href="https://archive.ics.uci.edu/ml/datasets/Census+Income" target="_blank">UCI Machine Learning Repository</a>. The datset was donated by Ron Kohavi and Barry Becker, after being published in the article <em>"Scaling Up the Accuracy of Naive-Bayes Classifiers: A Decision-Tree Hybrid"</em>. You can find the article by Ron Kohavi <a href="https://www.aaai.org/Papers/KDD/1996/KDD96-033.pdf" target="_blank">online</a>. The data we investigate here consists of small changes to the original dataset, such as removing the <code>'fnlwgt'</code> feature and records with missing or ill-formatted entries.</p>
</div></div>

<div>
 <h2>Finding Donors</h2>
<p>In this project, we will accurately model individuals' income using data collected from the 1994 U.S. Census. Our goal with this implementation is to construct a model that accurately predicts whether an individual makes more than $50,000. This sort of task can arise in a non-profit setting, where organizations survive on donations.  Understanding an individual's income can help a non-profit better understand how large of a donation to request, or whether or not they should reach out to begin with.  While it can be difficult to determine an individual's general income bracket directly from public sources, we can (as we will see) infer this value from other publically available features. </p>
<p>The dataset for this project originates from the <a href="https://archive.ics.uci.edu/ml/datasets/Census+Income" target="_blank">UCI Machine Learning Repository</a>.  </p>
</div>

In [1]:
import pandas as pd
import numpy as np
from time import time
import os
import re
os.listdir()

['.ipynb_checkpoints',
 'adult.data',
 'adult.names',
 'adult.test',
 'census.csv',
 'Finding_Donors.ipynb',
 'old.adult.names']

In [100]:
with open ('old.adult.names') as f:
    data=f.readlines()
col=[]
for i in data[64:78]:
    col_element=((re.findall('(\w*\-*\w*\-*\w+\:)', i)[0]).strip(':'))
    col.append(col_element)

col

['workclass',
 'fnlwgt',
 'education',
 'education-num',
 'marital-status',
 'occupation',
 'relationship',
 'race',
 'sex',
 'capital-gain',
 'capital-loss',
 'hours-per-week',
 'native-country',
 'class']

In [101]:
df_train=pd.read_csv('adult.data', header=None, names=col)


df_train.head()


workclass  fnlwgt   education  education-num       marital-status  \
39          State-gov   77516   Bachelors             13        Never-married   
50   Self-emp-not-inc   83311   Bachelors             13   Married-civ-spouse   
38            Private  215646     HS-grad              9             Divorced   
53            Private  234721        11th              7   Married-civ-spouse   
28            Private  338409   Bachelors             13   Married-civ-spouse   

            occupation    relationship    race      sex  capital-gain  \
39        Adm-clerical   Not-in-family   White     Male          2174   
50     Exec-managerial         Husband   White     Male             0   
38   Handlers-cleaners   Not-in-family   White     Male             0   
53   Handlers-cleaners         Husband   Black     Male             0   
28      Prof-specialty            Wife   Black   Female             0   

    capital-loss  hours-per-week  native-country   class  
39             0              40   United-States   <=50K  
50             0              13   United-States   <=50K  
38             0              40   United-States   <=50K  
53             0              40   United-States   <=50K  
28             0              40            Cuba   <=50K

In [102]:
def resume_table(df):
    print(f"Shape of the Dataframe: {df.shape}")
    summary=pd.DataFrame(data=df.dtypes, columns=['dtypes'])
    summary=summary.reset_index()
    summary['Name']=summary['index']
    summary=summary[['Name', 'dtypes']]
    summary['Missing']=df.isnull().sum().values
    summary['Uniques']=df.nunique().values
    
    return summary
resume_table(df_train)
    

Shape of the Dataframe: (32561, 14)


Name  dtypes  Missing  Uniques
0        workclass  object        0        9
1           fnlwgt   int64        0    21648
2        education  object        0       16
3    education-num   int64        0       16
4   marital-status  object        0        7
5       occupation  object        0       15
6     relationship  object        0        6
7             race  object        0        5
8              sex  object        0        2
9     capital-gain   int64        0      119
10    capital-loss   int64        0       92
11  hours-per-week   int64        0       94
12  native-country  object        0       42
13           class  object        0        2

In [103]:
np.round(df_train.describe(include='all'), 3).T

count unique                  top   freq     mean      std  \
workclass       32561      9              Private  22696      NaN      NaN   
fnlwgt          32561    NaN                  NaN    NaN   189778   105550   
education       32561     16              HS-grad  10501      NaN      NaN   
education-num   32561    NaN                  NaN    NaN   10.081    2.573   
marital-status  32561      7   Married-civ-spouse  14976      NaN      NaN   
occupation      32561     15       Prof-specialty   4140      NaN      NaN   
relationship    32561      6              Husband  13193      NaN      NaN   
race            32561      5                White  27816      NaN      NaN   
sex             32561      2                 Male  21790      NaN      NaN   
capital-gain    32561    NaN                  NaN    NaN  1077.65  7385.29   
capital-loss    32561    NaN                  NaN    NaN   87.304   402.96   
hours-per-week  32561    NaN                  NaN    NaN   40.437   12.347   
native-country  32561     42        United-States  29170      NaN      NaN   
class           32561      2                <=50K  24720      NaN      NaN   

                  min     25%     50%     75%          max  
workclass         NaN     NaN     NaN     NaN          NaN  
fnlwgt          12285  117827  178356  237051  1.48470e+06  
education         NaN     NaN     NaN     NaN          NaN  
education-num       1       9      10      12           16  
marital-status    NaN     NaN     NaN     NaN          NaN  
occupation        NaN     NaN     NaN     NaN          NaN  
relationship      NaN     NaN     NaN     NaN          NaN  
race              NaN     NaN     NaN     NaN          NaN  
sex               NaN     NaN     NaN     NaN          NaN  
capital-gain        0       0       0       0        99999  
capital-loss        0       0       0       0         4356  
hours-per-week      1      40      40      45           99  
native-country    NaN     NaN     NaN     NaN          NaN  
class             NaN     NaN     NaN     NaN          NaN

In [104]:
df_train['class'].value_counts()

 <=50K    24720
 >50K      7841
Name: class, dtype: int64

### Data Exploration Findings
Investigation of the dataset will determine how many individuals fit into either group, and will tell us about the percentage of these individuals making more than \$50,000. In the code cell below, you will need to compute the following:
- The total number of records, `'32561'`
- The number of individuals making more than \$50,000 annually: `7841`.
- The number of individuals making at most \$50,000 annually: `24720`.
- The percentage of individuals making more than \$50,000 annually: `24.1%`.



<h3> Featureset Exploration</h3>

* **age**: continuous. 
* **workclass**: Private, Self-emp-not-inc, Self-emp-inc, Federal-gov, Local-gov, State-gov, Without-pay, Never-worked. 
* **education**: Bachelors, Some-college, 11th, HS-grad, Prof-school, Assoc-acdm, Assoc-voc, 9th, 7th-8th, 12th, Masters, 1st-4th, 10th, Doctorate, 5th-6th, Preschool. 
* **education-num**: continuous. 
* **marital-status**: Married-civ-spouse, Divorced, Never-married, Separated, Widowed, Married-spouse-absent, Married-AF-spouse. 
* **occupation**: Tech-support, Craft-repair, Other-service, Sales, Exec-managerial, Prof-specialty, Handlers-cleaners, Machine-op-inspct, Adm-clerical, Farming-fishing, Transport-moving, Priv-house-serv, Protective-serv, Armed-Forces. 
* **relationship**: Wife, Own-child, Husband, Not-in-family, Other-relative, Unmarried. 
* **race**: Black, White, Asian-Pac-Islander, Amer-Indian-Eskimo, Other. 
* **sex**: Female, Male. 
* **capital-gain**: continuous. 
* **capital-loss**: continuous. 
* **hours-per-week**: continuous. 
* **native-country**: United-States, Cambodia, England, Puerto-Rico, Canada, Germany, Outlying-US(Guam-USVI-etc), India, Japan, Greece, South, China, Cuba, Iran, Honduras, Philippines, Italy, Poland, Jamaica, Vietnam, Mexico, Portugal, Ireland, France, Dominican-Republic, Laos, Ecuador, Taiwan, Haiti, Columbia, Hungary, Guatemala, Nicaragua, Scotland, Thailand, Yugoslavia, El-Salvador, Trinadad&Tobago, Peru, Hong, Holand-Netherlands.

## Preparing the Data
Before data can be used as input for machine learning algorithms, it often must be cleaned, formatted, and restructured — For this dataset, there are no invalid or missing entries we must deal with, however, there are some qualities about certain features that must be adjusted. This preprocessing can help tremendously with the outcome and predictive power of nearly all learning algorithms.

### Transforming Skewed Continuous Features
A dataset may sometimes contain at least one feature whose values tend to lie near a single number, but will also have a non-trivial number of vastly larger or smaller values than that single number.  Algorithms can be sensitive to such distributions of values and can underperform if the range is not properly normalized. With the census dataset two features fit this description: '`capital-gain'` and `'capital-loss'`. 



In [106]:
for i,value in enumerate(df_train.dtypes):
    if value !=object:
        print(df_train.columns[i])
    else:
        print(f'Object column: {df_train.columns[i]}')

Object column: workclass
fnlwgt
Object column: education
education-num
Object column: marital-status
Object column: occupation
Object column: relationship
Object column: race
Object column: sex
capital-gain
capital-loss
hours-per-week
Object column: native-country
Object column: class
